### PR TITLE
NeMo MoE docs

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: docs/source/features/memory_optimizations.rst [hunk 1]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
 Refer to :doc:`Parallelism <./parallelisms>`.
 
+
+Mixture of Experts
+------------------
+
+Overview
+^^^^^^^^
+
+NeMo supports Mixture of Experts (MoE)in the transformer layer for NLP models.
```
Comment (jgerh) [inline docs/source/features/memory_optimizations.rst:15]: Add space.

NeMo supports Mixture of Experts (MoE) in the transformer layer for NLP models.

File: docs/source/features/memory_optimizations.rst [hunk 2]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+
+To use MoE  in the NeMo Framework, adjust the ``num_moe_experts`` parameter in the model configuration:
+
+1. Set ``num_moe_experts`` to `8` to leverage 8 experts in the MoE module.
+
+   .. code-block:: yaml
+
+       num_moe_experts: 8  # Set MoE to use 8 experts
+
+2. Set ``moe_router_topk`` to the number of experts you want activated, for eaxmple if you want to process each input with two experts:
```
Comment (jgerh) [inline docs/source/features/memory_optimizations.rst:31]: Fix spelling abd add sentence.

2. Set ``moe_router_topk`` to the number of experts you want activated. For example, if you want to process each input with two experts:

File: docs/source/features/memory_optimizations.rst [hunk 3]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+
+   .. code-block:: yaml
+
+       moe_router_topk: 2  # Processes each token using 2 experts.
+
+In addition NeMo provides options to configure MoE-specific loss function, for example to balance token
```
Comment (jgerh) [inline docs/source/features/memory_optimizations.rst:37]: Fix punctuation and add sentence.

In addition, NeMo provides options to configure MoE-specific loss function. For example, to balance token
distribution across experts:

File: docs/source/features/memory_optimizations.rst [hunk 4]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+distribution across experts:
+
+1. Set ``moe_router_load_balancing_type`` to specify the load balancing method:
+
+   .. code-block:: yaml
+
+      moe_router_load_balancing_type: aux_loss  # to use the auxilary loss, other options include "sinkhorn".
+
+2. Set ``moe_aux_loss_coeff`` to specify the weight of the auxilary loss, values in the 1e-2 range are a good start:
```
Comment (jgerh) [inline docs/source/features/memory_optimizations.rst:46]: Fix punctuation and add sentence

2. Set ``moe_aux_loss_coeff`` to specify the weight of the auxilary loss. Values in the 1e-2 range are a good start.

File: docs/source/features/memory_optimizations.rst [hunk 5]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+
+   .. code-block:: yaml
+
+      moe_aux_loss_coeff: 1e-2  # set the aux-loss weight to 1e-2
+
+3. Set ``moe_z_loss_coeff`` to specify the weight of the z-loss, a starting value of 1e-3 is recommended:
```
Comment (jgerh) [inline docs/source/features/memory_optimizations.rst:52]: Fix punctuation and add sentence.

3. Set ``moe_z_loss_coeff`` to specify the weight of the z-loss.  A starting value of 1e-3 is recommended, as follows:

File: docs/source/features/memory_optimizations.rst [hunk 6]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+
+   .. code-block:: yaml
+
+      moe_z_loss_coeff: 1e-3
+
+Other options include:
+
+1. ``moe_input_jitter_eps`` adds noise to the input tensor by applying jitter with a specified epsilon value
```
Comment (jgerh) [inline docs/source/features/memory_optimizations.rst:60]: Add period.

1. ``moe_input_jitter_eps`` adds noise to the input tensor by applying jitter with a specified epsilon value.

File: docs/source/features/memory_optimizations.rst [hunk 7]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+4. ``moe_per_layer_logging`` enables per-layer logging for MoE, currently support aux-loss and z-loss.
+
+5. ``moe_expert_capacity_factor`` the capacity factor for each expert, None means no token will be dropped. The default is None.
+
+6. ``moe_pad_expert_input_to_capacity`` if True, pads the input for each expert to match the expert capacity length, effective only after the moe_expert_capacity_factor is set. The default setting is False.
+
+7. ``moe_token_drop_policy`` the policy to drop tokens. Can be either "probs" or "position". If "probs", the tokens with the lowest probabilities will be dropped. If "position", tokens at the end of each batch will be dropped. Default value is "probs".
+
+8. ``moe_layer_recompute`` if True, checkpointing moe_layer to save activation memory, default is False.
+
```
Comment (jgerh) [inline docs/source/features/memory_optimizations.rst:76]: Additional copyedits in the file.


**Line 181 Add period at end of sentence.**

Check implementation details from Attention Class in Megatron Core Repo: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/attention.py#L49.


**Lines 192 - 196. Bad formatting. The > character collapses the lines into a single paragraph. Remove > and replace with the rst bullet character *.**

* Supports training models with long sequence lengths by managing activation memory efficiently.
* Enables high batch sizes per GPU by offloading activation memory.
* Overlaps computation with data transfers (Host2Device and Device2Host) during offloading and reloading.

**Lines 200 - 202. Bad formatting. The > character collapses the lines into a single paragraph. Remove > and replace with the rst bullet character *..**

* Set cpu_offloading to True to enable CPU offloading.
* Set cpu_offloading_num_layers to a value between 0 and the total number of layers in the model minus one.
* Set cpu_offloading_activations and cpu_offloading_weights based on your needs to offload activations only, weights only, or both.

**Review by yaoyu-33 — APPROVED** [review]

File: docs/source/features/memory_optimizations.rst [hunk 8]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+       moe_router_topk: 2  # Processes each token using 2 experts.
+
+In addition, NeMo provides options to configure MoE-specific loss function.
+To balance token distribution across experts:
+
+1. Set ``moe_router_load_balancing_type`` to specify the load balancing method:
+
+   .. code-block:: yaml
+
+      moe_router_load_balancing_type: aux_loss  # to use the auxilary loss, other options include "sinkhorn".
```
Comment (ntajbakhsh) [inline docs/source/features/memory_optimizations.rst:44]: Can the user easily find the full list of options?

File: docs/source/features/memory_optimizations.rst [hunk 9]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
 Refer to :doc:`Parallelism <./parallelisms>`.
 
+
+Mixture of Experts
+------------------
+
+Overview
+^^^^^^^^
+
+NeMo supports Mixture of Experts (MoE) in the transformer layer for NLP models.
```
Comment (ntajbakhsh) [inline docs/source/features/memory_optimizations.rst:15]: It's not limited to NLP. Maybe we can just say "... in the feedforward block of a transformer layer."

File: docs/source/features/memory_optimizations.rst [hunk 10]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+In addition, NeMo provides options to configure MoE-specific loss function.
+To balance token distribution across experts:
+
+1. Set ``moe_router_load_balancing_type`` to specify the load balancing method:
+
+   .. code-block:: yaml
+
+      moe_router_load_balancing_type: aux_loss  # to use the auxilary loss, other options include "sinkhorn".
+
+2. Set ``moe_aux_loss_coeff`` to specify the weight of the auxilary loss. Values in the 1e-2 range are a good start, as follows:
```
Comment (ntajbakhsh) [inline docs/source/features/memory_optimizations.rst:46]: Does it make sense to briefly motive the need for the moe loss?

File: docs/source/features/memory_optimizations.rst [hunk 11]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+      moe_z_loss_coeff: 1e-3
+
+Other options include:
+
+1. ``moe_input_jitter_eps`` adds noise to the input tensor by applying jitter with a specified epsilon value.
+
+2. ``moe_token_dropping`` enables selectively dropping and padding tokens for each expert to achieve
+   a specified capacity.
+
+3. ``moe_token_dropping`` specifies the token dispatcher type, options include 'allgather' and 'alltoall'.
```
Comment (ntajbakhsh) [inline docs/source/features/memory_optimizations.rst:65]: This is the same as #2.

File: docs/source/features/memory_optimizations.rst [hunk 12]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+
+   .. code-block:: yaml
+
+      moe_z_loss_coeff: 1e-3
+
+Other options include:
+
+1. ``moe_input_jitter_eps`` adds noise to the input tensor by applying jitter with a specified epsilon value.
+
+2. ``moe_token_dropping`` enables selectively dropping and padding tokens for each expert to achieve
```
Comment (ntajbakhsh) [inline docs/source/features/memory_optimizations.rst:62]: Does it make sense to briefly motive the need for the token dropping?

File: docs/source/features/memory_optimizations.rst [hunk 13]
```diff
@@ -5,6 +5,80 @@ Parallelism
...
+   a specified capacity.
+
+3. ``moe_token_dropping`` specifies the token dispatcher type, options include 'allgather' and 'alltoall'.
+
+4. ``moe_per_layer_logging`` enables per-layer logging for MoE, currently support aux-loss and z-loss.
+
+5. ``moe_expert_capacity_factor`` the capacity factor for each expert, None means no token will be dropped. The default is None.
```
Comment (ntajbakhsh) [inline docs/source/features/memory_optimizations.rst:69]: Does it make sense to briefly motive the need for a capacity factor?

File: docs/source/features/memory_optimizations.rst (hunk 8 above)
Comment (akoumpa) [reply docs/source/features/memory_optimizations.rst:44]: this is the full list of options.
